src/transaction.rs

```diff
@@ -0,0 +1,554 @@
+    pub fn get_version(&self) -> Result<Version, Error> {
+        // get version from device
+        let response = APDU::new(YKPIV_INS_GET_VERSION).transmit(self, 261)?;
+
+        if !response.is_success() {
+            return Err(Error::GenericError);
+        }
+
+        if response.buffer().len() < 3 {
+            return Err(Error::SizeError);
+        }
+
+        Ok(Version {
+            major: response.buffer()[0],
+            minor: response.buffer()[1],
+            patch: response.buffer()[2],
+        })
+    }
+
+    /// Get YubiKey device serial number
+    pub fn get_serial(&self, version: Version) -> Result<Serial, Error> {
+        let yk_applet = [0xa0, 0x00, 0x00, 0x05, 0x27, 0x20, 0x01, 0x01];
+
+        let response = if version.major < 5 {
+            // get serial from neo/yk4 devices using the otp applet
+            let sw = APDU::new(YKPIV_INS_SELECT_APPLICATION)
+                .p1(0x04)
+                .data(&yk_applet)
+                .transmit(self, 0xFF)?
+                .status_words();
+
+            if !sw.is_success() {
+                error!("failed selecting yk application: {:04x}", sw.code());
+                return Err(Error::GenericError);
+            }
+
+            let resp = APDU::new(0x01).p1(0x10).transmit(self, 0xFF)?;
+
+            if !resp.is_success() {
+                error!(
+                    "failed retrieving serial number: {:04x}",
+                    resp.status_words().code()
+                );
+                return Err(Error::GenericError);
+            }
+
+            // reselect the PIV applet
+            let sw = APDU::new(YKPIV_INS_SELECT_APPLICATION)
+                .p1(0x04)
+                .data(&AID)
+                .transmit(self, 0xFF)?
+                .status_words();
+
+            if !sw.is_success() {
+                error!("failed selecting application: {:04x}", sw.code());
+                return Err(Error::GenericError);
+            }
+
+            resp
+        } else {
+            // get serial from yk5 and later devices using the f8 command
+            let resp = APDU::new(YKPIV_INS_GET_SERIAL).transmit(self, 0xFF)?;
+
+            if !resp.is_success() {
+                error!(
+                    "failed retrieving serial number: {:04x}",
+                    resp.status_words().code()
+                );
+                return Err(Error::GenericError);
+            }
+
+            resp
+        };
+
+        response.buffer()[..4]
+            .try_into()
+            .map(|serial| Serial::from(u32::from_be_bytes(serial)))
+            .map_err(|_| Error::SizeError)
+    }
+
+    /// Verify device PIN.
+    pub fn verify_pin(&self, pin: &[u8]) -> Result<(), Error> {
+        if pin.len() > CB_PIN_MAX {
+            return Err(Error::SizeError);
+        }
+
+        let response = APDU::new(YKPIV_INS_VERIFY)
+            .params(0x00, 0x80)
+            .data(pin)
```

Comment:
This function is not padding `pin` with `0xff`, and is not verifying that `pin.len() == CB_PIN_MAX` (placing the onus on the caller to pad). One or the other should be done.

Reply:
FWIW, the logic is identical to the `_verify` function from `ykpiv.c`, but it seems like we should be able to change the conditional above to be `if pin.len() != CB_PIN_MAX`.

Not sure how much the padding matters or if it could parse a shorter PIN from the APDU...

Reply:
Added a stricter check in c54f66a